File: app/runner.py
"""HTTP 任务执行引擎与评估逻辑模块。

处理任务的 HTTP 请求发起、响应结果关键词评估、并发控制、抢零点/Burst重试模式以及历史记录保存。
"""

import asyncio
import time
from typing import List

import httpx

from app.config import format_now, format_now_ms
from app.models import HistoryItem, Task
from app.notifier import send_notifications
from app.storage import StateManager

# 高性能连接池配置
_limits = httpx.Limits(max_keepalive_connections=128, max_connections=64)
_client = httpx.AsyncClient(limits=_limits, follow_redirects=True)

SCHEDULED_BURST_RETRIES: int = 3
SCHEDULED_BURST_INTERVAL: float = 0.250  # 250 毫秒


def evaluate_response(task: Task, status_code: int, body: str) -> str:
    """评估 HTTP 响应状态码及返回体关键词。

    根据 HTTP Status Code 和任务配置的成功/失败关键词判断请求判定结果。
    """
    if status_code < 200 or status_code >= 300:
        return f"http {status_code}"

    body_lower = body.lower()

    # 校验失败关键词
    if task.failure_keywords:
        fail_words = [w.strip().lower() for w in task.failure_keywords.split(",") if w.strip()]
        for word in fail_words:
            if word in body_lower:
                return "failure keyword matched"

    # 校验成功关键词
    if task.success_keywords:
        succ_words = [w.strip().lower() for w in task.success_keywords.split(",") if w.strip()]
        if succ_words:
            matched = any(word in body_lower for word in succ_words)
            if not matched:
                return "success keyword missing"

    return "success"


def should_use_burst_mode(task: Task, triggered_by: str) -> bool:
    """判断是否应当开启抢零点 Burst 重试模式。"""
    if not task.aggressive_mode:
        return False
    return triggered_by in ("schedule", "manual-schedule-check")


async def execute_task(task: Task, triggered_by: str, state_manager: StateManager) -> HistoryItem:
    """执行单个签到任务。

    包含最大重试、抢零点模式逻辑、结果评估与历史保存。
    """
    max_attempts: int = task.retry_count + 1
    burst_mode: bool = should_use_burst_mode(task, triggered_by)

    if burst_mode and max_attempts < SCHEDULED_BURST_RETRIES + 1:
        max_attempts = SCHEDULED_BURST_RETRIES + 1

    last_message: str = ""
    last_status_code: int = 0
    preview: str = ""
    duration_ms: int = 0
    last_request_started_at: str = ""

    for attempt in range(max_attempts):
        started_time: float = time.time()
        request_started_at: str = format_now_ms()
        last_request_started_at = request_started_at

        try:
            req_headers = dict(task.headers)
            body_bytes = task.body.encode("utf-8") if task.body else None
            response = await _client.request(
                method=task.method,
                url=task.url,
                headers=req_headers,
                content=body_bytes,
                timeout=float(task.timeout_seconds),
            )

            duration_ms = int((time.time() - started_time) * 1000)
            last_status_code = response.status_code

            # 截取前 4096 字节作为预览
            preview_bytes = response.content[:4096]
            preview = preview_bytes.decode("utf-8", errors="replace")

            last_message = evaluate_response(task, response.status_code, preview)

            if last_message == "success":
                result = HistoryItem(
                    task_id=task.id,
                    task_name=task.name,
                    status="success",
                    status_code=response.status_code,
                    message="request completed",
                    response_preview=preview,
                    response_time_ms=duration_ms,
                    request_started_at=request_started_at,
                    triggered_by=triggered_by,
                    created_at=format_now(),
                )
                await finish_run(task.id, result, state_manager)
                return result

        except Exception as exc:
            duration_ms = int((time.time() - started_time) * 1000)
            last_message = str(exc)

        # 失败重试等待
        if attempt < max_attempts - 1:
            if burst_mode and attempt < SCHEDULED_BURST_RETRIES:
                await asyncio.sleep(SCHEDULED_BURST_INTERVAL)
            else:
                await asyncio.sleep(attempt + 1)

    result = HistoryItem(
        task_id=task.id,
        task_name=task.name,
        status="failed",
        status_code=last_status_code,
        message=last_message,
        response_preview=preview,
        response_time_ms=duration_ms,
        request_started_at=last_request_started_at,
        triggered_by=triggered_by,
        created_at=format_now(),
    )
    await finish_run(task.id, result, state_manager)
    return result


async def finish_run(task_id: int, item: HistoryItem, state_manager: StateManager) -> None:
    """完成任务运行，记录历史日志并写回磁盘，最后异步发送通知。"""
    async with state_manager._lock:
        item.id = state_manager.state.next_history_id
        state_manager.state.next_history_id += 1
        state_manager.state.history.insert(0, item)

        # 限制历史记录最多 200 条
        if len(state_manager.state.history) > 200:
            state_manager.state.history = state_manager.state.history[:200]

        for t in state_manager.state.tasks:
            if t.id == task_id:
                t.last_status = item.status
                t.last_run_at = item.created_at
                t.last_duration_ms = item.response_time_ms
                t.updated_at = format_now()
                break

        await state_manager._save_unlocked()

    # 异步推送通知
    asyncio.create_task(send_notifications(item, state_manager.state.settings.notify))


async def run_all_enabled_tasks(
    triggered_by: str, state_manager: StateManager
) -> List[HistoryItem]:
    """并发批量运行所有启用的任务。"""
    tasks = [t for t in state_manager.state.tasks if t.enabled]
    if not tasks:
        return []

    max_workers = state_manager.state.settings.schedule.max_workers
    if max_workers < 1:
        max_workers = 4
    elif max_workers > 8:
        max_workers = 8

    sem = asyncio.Semaphore(max_workers)

    async def worker(task: Task) -> HistoryItem:
        async with sem:
            return await execute_task(task, triggered_by, state_manager)

    results = await asyncio.gather(*(worker(t) for t in tasks))
    # 按 task_id 倒序排列
    sorted_results = sorted(list(results), key=lambda x: x.task_id, reverse=True)
    return sorted_results

File: runner.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"io"
	"net/http"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	scheduledBurstRetries  = 3
	scheduledBurstInterval = 250 * time.Millisecond
)

var defaultTransport = &http.Transport{
	MaxIdleConns:        128,
	MaxIdleConnsPerHost: 64,
	IdleConnTimeout:     90 * time.Second,
}

var runnerHTTPClient = &http.Client{
	Transport: defaultTransport,
}

// EvaluateResponse 评估 HTTP 状态码及关键字规则
func EvaluateResponse(task Task, statusCode int, body string) string {
	if statusCode < 200 || statusCode >= 300 {
		return fmt.Sprintf("http %d", statusCode)
	}

	bodyLower := strings.ToLower(body)

	// 校验失败关键字
	if task.FailureKeywords != "" {
		failWords := strings.Split(task.FailureKeywords, ",")
		for _, w := range failWords {
			word := strings.ToLower(strings.TrimSpace(w))
			if word != "" && strings.Contains(bodyLower, word) {
				return "failure keyword matched"
			}
		}
	}

	// 校验成功关键字
	if task.SuccessKeywords != "" {
		succWords := strings.Split(task.SuccessKeywords, ",")
		var matched bool
		hasValidWord := false
		for _, w := range succWords {
			word := strings.ToLower(strings.TrimSpace(w))
			if word != "" {
				hasValidWord = true
				if strings.Contains(bodyLower, word) {
					matched = true
					break
				}
			}
		}
		if hasValidWord && !matched {
			return "success keyword missing"
		}
	}

	return "success"
}

// ShouldUseBurstMode 判断是否开启抢零点 Burst 重试模式
func ShouldUseBurstMode(task Task, triggeredBy string) bool {
	if !task.AggressiveMode {
		return false
	}
	return triggeredBy == "schedule" || triggeredBy == "manual-schedule-check"
}

// ExecuteTask 执行单个签到任务
func ExecuteTask(ctx context.Context, task Task, triggeredBy string, stateMgr *StateManager) HistoryItem {
	maxAttempts := task.RetryCount + 1
	burstMode := ShouldUseBurstMode(task, triggeredBy)

	if burstMode && maxAttempts < scheduledBurstRetries+1 {
		maxAttempts = scheduledBurstRetries + 1
	}

	var lastMessage string
	var lastStatusCode int
	var preview string
	var durationMS int
	var lastRequestStartedAt string

	for attempt := 0; attempt < maxAttempts; attempt++ {
		startTime := time.Now()
		requestStartedAt := FormatNowMS()
		lastRequestStartedAt = requestStartedAt

		reqTimeout := time.Duration(task.TimeoutSeconds) * time.Second
		if reqTimeout <= 0 {
			reqTimeout = 30 * time.Second
		}
		reqCtx, cancel := context.WithTimeout(ctx, reqTimeout)

		var reqBody io.Reader
		if task.Body != "" {
			reqBody = bytes.NewBufferString(task.Body)
		}

		req, err := http.NewRequestWithContext(reqCtx, task.Method, task.URL, reqBody)
		if err == nil {
			for k, v := range task.Headers {
				req.Header.Set(k, v)
			}

			resp, err := runnerHTTPClient.Do(req)
			if err == nil {
				durationMS = int(time.Since(startTime).Milliseconds())
				lastStatusCode = resp.StatusCode

				// 最多读取前 4096 字节
				bodyBytes, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
				_ = resp.Body.Close()
				preview = string(bodyBytes)

				lastMessage = EvaluateResponse(task, resp.StatusCode, preview)
				cancel()

				if lastMessage == "success" {
					result := HistoryItem{
						TaskID:           task.ID,
						TaskName:         task.Name,
						Status:           "success",
						StatusCode:       resp.StatusCode,
						Message:          "request completed",
						ResponsePreview:  preview,
						ResponseTimeMS:   durationMS,
						RequestStartedAt: requestStartedAt,
						TriggeredBy:      triggeredBy,
						CreatedAt:        FormatNow(),
					}
					finishRun(task.ID, result, stateMgr)
					return result
				}
			} else {
				durationMS = int(time.Since(startTime).Milliseconds())
				lastMessage = err.Error()
				cancel()
			}
		} else {
			lastMessage = err.Error()
			cancel()
		}

		// 失败重试等待
		if attempt < maxAttempts-1 {
			if burstMode && attempt < scheduledBurstRetries {
				time.Sleep(scheduledBurstInterval)
			} else {
				time.Sleep(time.Duration(attempt+1) * time.Second)
			}
		}
	}

	result := HistoryItem{
		TaskID:           task.ID,
		TaskName:         task.Name,
		Status:           "failed",
		StatusCode:       lastStatusCode,
		Message:          lastMessage,
		ResponsePreview:  preview,
		ResponseTimeMS:   durationMS,
		RequestStartedAt: lastRequestStartedAt,
		TriggeredBy:      triggeredBy,
		CreatedAt:        FormatNow(),
	}
	finishRun(task.ID, result, stateMgr)
	return result
}

// finishRun 保存记录、更新任务运行状态并异步推送通知
func finishRun(taskID int, item HistoryItem, stateMgr *StateManager) {
	stateMgr.Mu.Lock()

	item.ID = stateMgr.State.NextHistoryID
	stateMgr.State.NextHistoryID++

	// 插入到最前面
	stateMgr.State.History = append([]HistoryItem{item}, stateMgr.State.History...)

	// 限制历史记录最多 200 条
	if len(stateMgr.State.History) > 200 {
		stateMgr.State.History = stateMgr.State.History[:200]
	}

	for i := range stateMgr.State.Tasks {
		if stateMgr.State.Tasks[i].ID == taskID {
			stateMgr.State.Tasks[i].LastStatus = item.Status
			stateMgr.State.Tasks[i].LastRunAt = item.CreatedAt
			stateMgr.State.Tasks[i].LastDurationMS = item.ResponseTimeMS
			stateMgr.State.Tasks[i].UpdatedAt = FormatNow()
			break
		}
	}

	notifySettings := stateMgr.State.Settings.Notify
	_ = stateMgr.saveUnlocked()
	stateMgr.Mu.Unlock()

	// 异步推送通知
	go SendNotifications(item, notifySettings)
}

// RunAllEnabledTasks 并发运行所有已启用的任务
func RunAllEnabledTasks(ctx context.Context, triggeredBy string, stateMgr *StateManager) []HistoryItem {
	stateMgr.Mu.RLock()
	var enabledTasks []Task
	for _, t := range stateMgr.State.Tasks {
		if t.Enabled {
			enabledTasks = append(enabledTasks, t)
		}
	}
	maxWorkers := stateMgr.State.Settings.Schedule.MaxWorkers
	stateMgr.Mu.RUnlock()

	if len(enabledTasks) == 0 {
		return []HistoryItem{}
	}

	if maxWorkers < 1 {
		maxWorkers = 4
	} else if maxWorkers > 8 {
		maxWorkers = 8
	}

	sem := make(chan struct{}, maxWorkers)
	resultsChan := make(chan HistoryItem, len(enabledTasks))
	var wg sync.WaitGroup

	for _, t := range enabledTasks {
		wg.Add(1)
		go func(task Task) {
			defer wg.Done()
			sem <- struct{}{}
			res := ExecuteTask(ctx, task, triggeredBy, stateMgr)
			<-sem
			resultsChan <- res
		}(t)
	}

	wg.Wait()
	close(resultsChan)

	var results []HistoryItem
	for r := range resultsChan {
		results = append(results, r)
	}

	// 按 TaskID 倒序排序
	sort.Slice(results, func(i, j int) bool {
		return results[i].TaskID > results[j].TaskID
	})

	return results
}
